### ISSUE TITLE
[Feature] Custom sockets implementation interaface

### ISSUE THREAD

**Issue by Faless** [warmcat/libwebsockets#1216]
As I mentioned before we integrated libwebsockets inside Godot Engine.
Godot Engine has its own platform independent socket implementation. and I was thinking it would be nice if lws implemented a way for developers to use custom sockets operation in place of the platform ones.

The idea is to provide something similar to `fops` (if I understand it correctly) for all the sockets operation (from creation to polling/reading/writing/etc) and allow to configure a custom user pointer to be passed to that function (so to allow the application to have context).

I had a quick look at the code and it does sound a bit tricky actually (many of those ops are sparse in the main library files) but I think it would both help the developers who integrate the library and allow for a better separation of the code (less `#ifdef` to check if it's Unix/Windows/etc).

Would that something you are interested in?

**Comment by lws-team** [comment]
I'm not sure I understood the proposal correctly...

The wsi itself wraps most details about the connection, whether it is tcp or udp, inside tls, has http 1.x, http/2 or ws, or is a RAW, user-protocol only socket, or even file handle, and has information about protocol state.

In the end every wsi has to participate in an event loop, poll() or one of the event loop libraries, which mandates it must be hung on a file or socket handle.  lws already supports all the variations of that.

The fops VFS stuff is about hiding details of the underlying storage when it radically differs, for example still use posix-style open() read() etc when it's the OS VFS as normal, or actually a ROMFS in flash and / or being served compressed out of a zip file (both used on ESP32).

I don't know what the "platform independent socket library" does but in the end it should boil down to the same socket(), recv() etc posix apis we use directly (which are pretty platform independent already, eg, work on ESP32 too).  What does it buy you if lws goes through that abstraction layer in your library too?

**Comment by Faless** [comment]
I see multiple advantages from an integration point of view:

- less network code overall, both in size and to scrutiny / maintain
(ipv4/ipv6 behavior, address resolution, ssl, socket options etc)
- consistent socket abstraction in all the libraries (we are using this
approach with the ENet library integration for example)
- ability to use our own SSL abstraction without having to include
libwebsockets' one.
- ability to use different protocols without hacking the library itself
(though this is not our case).

**Comment by lws-team** [comment]
I think adding two extra layers (a generic one, and then the ENet wrappers) just to call some posix libc api we already call directly 

https://github.com/lsalzman/enet/blob/master/unix.c#L295

doesn't make sense for us.  The clashing event loops and other sundries like retry timers will also be messy to integrate even if you had that.

I would be much more interested to help with porting ENet to use lws natively, either as a protocol plugin, or expose build-optional apis or some mixture. So it natively uses the lws event loop, existing lws apis for UDP socket generation and adoption, lws apis for timer, and probably lws callback for RX and udp sockaddr capture.  That may sound a bit of a strange counter-proposal, but ENet is BSD licensed and very mature, there have been few patches on it since 2014.  Eyeballing the related code in git briefly it seems simple enough to consider it.

In fact more than ENet I am already very interested in Quic, which is a superset of ENet and "useful for other things" in a clear way.  The recent stuff in lws around udp and hrtimer is aimed at starting on Quic shortly.

**Comment by Faless** [comment]
Well, I mentioned ENet not for integrating it in lws but as an example of how we did the integration of it inside Godot Engine, using our own socket implementation eg:

https://github.com/godotengine/godot/blob/master/thirdparty/enet/godot.cpp#L93-L106 or https://github.com/godotengine/godot/blob/master/thirdparty/enet/godot.cpp#L123-162 (see that the library is using our own `PacketPeerUDP` abstraction and we do not include `unix.c` at all).

**Comment by lws-team** [comment]
For lws, this change doesn't buy us anything... every platform we support (every platform I know of... even grungy RTOSes) provide posix socket apis.  There is no use in lws for an abstraction that only ever resolves to what we already have directly.

You listed some "advantages" earlier, but none of these advantages apply to lws, which is where you want to do the work.  Quite the opposite with "less network code overall" from lws POV.  Since I maintain lws, that leaves me a bit cold.  If I missed the point I will be happy to see the error of my ways, but please consider this from lws maintainer POV rather than your project POV first.

**Comment by Faless** [comment]
@lws-team I understand your point of view, and I agree that it might end up being more complex from the library POV.
But lws is, as said, a **library** (and a great one!), which means it is supposed to be used by other apps.

I didn't argue before, because I didn't have a direct use case to bring to you, but yesterday I got one.
As mentioned before we use lws in Godot Engine, which is a multi platform game engine. One of the devs maintains console ports (like PlayStation and Nintendo). Sadly, many of those platform **do not come with BSD sockets, nor Winsock**. Which results in the library not compiling, and to fix that, we need to hack the library itself (sadly, their SDKs are NDAed bullshit, so I have no link for you, and they can't be supported in the library itself).
Having a `sockops` struct that we can define to use for socket operation will allow applications to support those exotic platforms without having to hack the library itself (like with `fops`).
Again, I understand this might be very low priority from your POV (or even useless), but it will make life for applications that uses the library much easier when porting to non POSIX/Winsocks environment.

**Comment by lws-team** [comment]
Well, since we last discussed it I start to implement QUIC.  Basically it moves all the packet generation responsibility into userland, including for retransmission and logical routing.  At that point, the application can also be extended to drive the NIC or receive and issue the packets by dataplane rather than go via the kernel.  There would not be any posix sockets then either.

What kind of set of ops is needed for these consoles?  In the dataplane case at least, it seems a related issue is it wants to allocate tx buffer memory from its own pools... the lifecycle of that vs retransmits etc needs consideration as well.  For zerocopy, lws would need a new api to acquire the tx buffer address in the user code somehow.  In QUIC, there is no longer kernel tcp window management providing POLLOUT.  In rx case, the same process will handle packet reception, ie no poll().  How do the consoles signal network events?

Also, a single device may have traditional kernel network interfaces with posix sockets, and eg 10G interfaces managed by userland dataplane.  To allow it, it impiles a per-wsi sockops pointer chosen by which interface it was bound to.  So it needs to support multiple "lws_sock_ops" active in the same process (All this doesn't have to work at once, but the architectural implications need to be thought about from the start).

The affected areas are at least

 - abstractions for POSIX sockets
 - buffer acquisition and an api user code can get the address
 - network event wait / signalling

Actually for the event wait issue, this abstraction idea doesn't change that the user will still require a single wait and event loop (eg, libuv) support, even when his event loop doesn't understand (some of) his networking arrangements and cannot wait on his connections.  He may be using the event loop to wait on device nodes, files, or maybe even mix-and-match between posix and other sockets.  If the non-posix stuff is handled asynchronously by interrupt or other thread, it can use a pipe the same way cancel_service does to wake the event loop / poll wait cheaply.

**Comment by punto-** [comment]
As Faless said I can't go into detail about these APIs, but they're usually not that different from Winsock, which is like BSD but without tying the sockets into a low-level "everything is a socket" OS abstraction. So for example winsock has "closesocket()" instead of "close()", because close() works on all first-class-sockets from the OS, and the winsock sockets are something else. There are other reasons for the host program to provide its own network layer (and memory allocation), for example if they're using something other than TCP (websockets over carrier pigeon?), or if they need to monitor the data that goes in and out for some reason, like statistics, etc.
We don't really need anything too fancy, other than to provide a read and write function, and memory allocation, similar to mbedtls for example ( https://tls.mbed.org/api/ssl_8h.html#a8b7442420aef7f1a76fa8c5336362f9e )

**Comment by lws-team** [comment]
So how do they perform event waits, that can eg, also wait on a file like poll() can?

**Comment by lws-team** [comment]
The windows WSA stuff is handled using lws platform abstraction...

https://github.com/warmcat/libwebsockets/blob/master/lib/plat/lws-plat-win.c

If it's actually messed-up-posix-by-another-name then the platform apis are probably enough, plus or minus some beautification.

"What you need" is only part of the problem, you need to share enough general information that we can figure out how to both get something useful and maintainable from it.

**Comment by punto-** [comment]
I assume they use the _timeout version of the function with infinite time to wait (we never block on sockets so I don't know in detail, we pass NULL on that parameter). Maybe that doesn't cover 100% this case, other example is Enet as Faless pointed out

"messed up posix" is a good way to describe these APIs generally

**Comment by punto-** [comment]
I'm not sure I understand the question.. You don't know what is behind that interfase, it could be a socket or a file or a stream or whatever. If they provide a _wait version, you use it and you expect it to wait. It's not your problem anymore if it's a socket or not, you just have something that feeds data into the protocol level logic, and takes data out..
I the case of mbedtls I think set_bio call is only used for sockets, and they have something else for files, but if you decide that it's for any io object, that's fine (as long as we can provide a non-blocking version too)

**Comment by lws-team** [comment]
As I said above we need to do something about

 - abstractions for POSIX sockets

It has a socket handle concept like WSA?  It has accept(), recvfrom(), sendto() etc or equivalents?  If they're direct equivalents, you can wrap them with a posix-compatible api name?

 - buffer acquisition and an api user code can get the address

Where do these apis want their buffer to have come from if, eg, you are sending something?  There is a special requirement to have allocated it from a platform api, or on the stack / malloc is OK?  Even for UDP-related apis?

 - network event wait / signalling

Lws expects to be able to use a single wait on a set of descriptors that may point to files or sockets or whatever.

So if you have a raw file wsi pointing to a file like /dev/input/... that will wake it as much as a socket becoming writable.  You can mix and match potentially thousands of those.

If no events, then it will block in the wait at 0% cpu (it must wake at 1hz for timeouts etc).

I get it that's not what you will do, but we need the two worlds to meet up.  If the platforms can do that we can figure it out.

**Comment by punto-** [comment]
I don't think the websocket layer needs to know about "connect" and "accept" and "bind" and DNS resolution. Let the network layer do that. The websocket just needs a pipe where data comes in and another pipe where data goes out. Maybe it needs to know the hostname (if it's still like in http/1.1), so you give a string with it. All that network stuff should be hidden behind those 2 pipes.

**Comment by lws-team** [comment]
@punto-  if that's what you think you want, I'm sure you can roll your own from scratch in an afternoon and since there's no basis to cooperate then, no need to bother me about it.

@Faless what you're describing is the current plat way of integrating "broken sockets", that windows uses.  There are already wrapper types for these things established.

For the case of the wait though, lws already supports libuv, libevent and libev event loops, not just poll; the event loop stuff already has its own ops abstraction.  The plat code just needs to implement the default service api using whatever wait it has in code; the user can choose to use that or an event loop.

> Then, the library could call those functions when needed (e.g. in service, when creating, etc) instead of doing socket handling directly 

It's already like that, with the higher level "ops" provided by plat stuff atm.  There's no point with a predefined list of abstracted socket options, there's one correct set (which may not all be bool) for a given platform and the platform code knows what it is.  Lws non-plat doesn't call setsockopt piecemeal but a plat api that "sets up a socket" for the platform.  Likewise lws doesn't use bind(), it calls through to plat code to "create a listen wsi", which uses bind() or whatever from plat code.

The only problem with saying just use plat is the platform choice implies the networking api choice atm.  In fact even with your abstraction layer, that is still true - on windows your abstraction layer will also resolve to WPA, on linux: posix, making it a NOP.

As I said earlier the only reason I would implement literally socket ops is related to QUIC / SDR stuff, that is kind of interesting atm.  However I need to research the details further.  Plat stuff is overdue a refactor anyway.
